Create Bulk Endpoint for Workspaces

Mythicaeda opened the pull request:
* **Tickets addressed:** Closes #1749 
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->
**Note: depends on #1757** 

Adds in bulk endpoints for file/directory level `PUT`, `POST`, and `DELETE` requests.

Updates single-item `PUT` to accept `folder` as a valid value for `type`. It is treated as equivalent to `directory`.

Refactors `handleMove` and `handleCopy`. They now take in all the values they would extract from `context` and output a `results` object that contains the HTTP status code and response object. This allows the bulk version of `POST` to call these methods in a loop and create its response array from the output.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
Basic E2E Tests were created. Edge case E2E Tests need to be created

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
No docs were invalidated.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
- Create API docs for Workspace Server
- Finish Workspace Binding E2E tests
- Concurrent File Operation Protections

Mythicaeda committed:
Reorder file and add "region" tags

Allows using the "structure" view to more easily find specific methods

Mythicaeda committed:
Implement bulk endpoints

Mythicaeda committed:
Rename "filePath" param to more generic "path"

Mythicaeda committed:
Use new type "ItemType" in single-item "put"

Mythicaeda committed:
Forbid providing both "moveTo" and "copyTo" in single-item POST

- fix minor errorMSG bug in handleMove
- update helptext in POST method to be stylistically consistent with remaining help texts

Mythicaeda committed:
Make bulkPost and Post behavior align

Rather than having slightly different methods for the bulk and non-bulk move/copy behavior, the non-bulk version was modified so that the bulk version could call it and ingest the result.

Mythicaeda committed:
Make delete directory recursive

Before, the server would not delete directories that had things in them

Mythicaeda committed:
Fix bulk move and bulk copy

- Fix destination path to cut out intermediate directories (like in single move/copy)

Mythicaeda committed:
Add Basic Bulk E2E Tests

- updates playwright in order to use formData.append

Mythicaeda committed:
Ensure that uploaded files have unique names in bulk PUT

Mythicaeda committed:
Fix bug preventing file overwrite in move and copy

Mythicaeda committed:
Fix PUT multipart check

Mythicaeda committed:
E2E Tests for Bulk POST Overwrite

Mythicaeda committed:
Report SecurityException if thrown

Mythicaeda committed:
Rename 'BulkPut' to 'BulkUpload'

Mythicaeda committed:
Update input body for BulkPost

Now allows individual items to be renamed copied/moved and placed under a different name

Mythicaeda committed:
Fix whitespace PostBody

Mythicaeda committed:
Fix metadata files not being deleted

Mythicaeda committed:
BulkPut/BulkPost - ensure that all destination paths are unique

Mythicaeda committed:
Update Workspace E2E tests to reflect Post input change

Mythicaeda committed:
Specify MALFORMED_REQUEST instead of INTERNAL_ERROR when the request body has issues

Mythicaeda committed:
Fix incorrect return structure in bulkUpload

Mythicaeda committed:
Ensure that the user has specified items to alter in BulkPost

Mythicaeda committed:
Finish Bulk E2E Tests

Now checks Malformed Requests, additional optional flags, and mixed-response codes within the 207

Mythicaeda committed:
Validate file and folder names to avoid illegal characters/reserved filenames

Mythicaeda committed:
Check for sibling directory cases in BulkTests

dandelany commented on workspace-server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -108,8 +127,17 @@ public void apply(final Javalin javalin) {
+      logger.error("Unexpected error processing workspace request", ex);
+      final var message = ex.getMessage() != null ? ex.getMessage() : "Unknown error.";
+      ctx.status(500).json(new FormattedError("UNKNOWN_ERROR", message, ex));
+    });
```
I did some testing on error cases and was able to hit this catch-all a couple times - good addition 👍 A few small recommendations:
* It's fine to not log noisy user problems like 404's, but we should add `logger.warn` or `logger.error` to handlers which indicate an "us problem", to help us debug what happened. Eg. at least `IOException` & `SQLException` currently return 500 but don't log.  Maybe `SecurityException` too since it either indicates a malicious user or a script gone wrong.
* Minor, but Javalin will sometimes throw specific HTTP exceptions which have already set HTTP status codes, and the catch-all handler will overwrite these as 500. I couldn't come up with a repro case without commenting out other handlers, but I think something like this would be a good defensive thing to put at the top of the catch-all handler:
```
  if (ex instanceof HttpResponseException hre) {
    ctx.status(hre.getStatus())
       .json(new FormattedError(hre));
    return;
  }
  ```

Mythicaeda replied:
Added a commit to address this: [Log when a 500 is returned by the server](https://github.com/NASA-AMMOS/aerie/pull/1760/commits/bec83d3c5e1412062802a4aa29d82b539aaa8559)

dandelany commented on workspace-server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -144,6 +172,37 @@ private JWTService.UserSession authorize(Context context) {
+   * If it does not, format the context into an appropriate error state.
+   * @return true, if the user passes the permissions check. false otherwise
+   */
+  private boolean checkPermissions(Context context, int workspaceId, WorkspaceAction action) {
```
No action needed, but i wonder if a better pattern here could be to `throw` on failed permission checks rather than mutating context and returning a boolean. That would centralize error handling, keep this helper more pure, and avoid callers needing to remember to `return` after the `false` case.

Mythicaeda replied:
I made this method a `boolean` that mutates the context to avoid having duplicate `catch` blocks everywhere that calls the method, but I suppose having it throw an exception that contains the status code and formatted error would work (though that wouldn't necessarily remove the need for a return statement, unless you're proposing to have none of the endpoints catch the new exception)

Mythicaeda replied:
As discussed, I'm going to make a separate small PR that puts that exception into the PermissionsService and updates `MerlinBindings`, `SchedulerBindings`, and `WorkspaceBindings` accordingly.

Edit: [here's the ticket for that work](https://github.com/NASA-AMMOS/aerie/issues/1771)

dandelany commented on workspace-server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -608,40 +654,464 @@ private void delete(Context context) throws NoSuchWorkspaceException, IOExceptio
+            response.add("status", 500)
+                    .add("response", new FormattedError("Could not save file.").toJson());
+          }
+        } catch (IOException ioe) {
```
There is a decent amount of duplicated logic between this function and the single-upload `put` handler... Might be nice to abstract them into a single `handleUpload` like you've done for `handleCopy` and `handleMove`. Or if that feels too big - it would be good to compare this side-by-side with that function and make sure they handle the same cases - eg. this handles `IOException` and `"Unsupported item upload type"` but I don't think the single `put` does.

Mythicaeda replied:
Addressed in this commit: [Extract Duplicate Logic for Upload and Delete](https://github.com/NASA-AMMOS/aerie/pull/1760/commits/9542e744a19135198785d205ba532c4341889623)

dandelany commented on e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/BindingsTests.java:
```diff
@@ -1484,7 +1486,7 @@ void nonEmptyWorkspaceWithDepth() {
 
     /**
-     * Tests for the /ws/{workspaceId}/<filePath> routes.
+     * Tests for the /ws/{workspaceId}/<path> routes.
```
Noticed above on line 1276 (couldn't leave a comment there since not a new change, but figured worth mentioning) - I'm a bit surprised that calling `deleteMockCommandDictionary` *before* `deleteMockParcel` doesn't throw/complain, since I thought the parcel had a foreign key for the command dictionary. If I try to delete a dictionary before deleting its parcel via UI, I get an error.

Mythicaeda replied:
So the reason it works is because `deleteParcel` is effectively doing nothing. The db relationship on `parcel` says that if it's `command_dictionary` is deleted, that delete should cascade and delete the `parcel` as well. I just tested in the UI and this worked as I said. In your UI test, was the parcel in use on a workspace? If so, the workspace would stop the parcel from being deleted, which would stop the cdict from being deleted.

dandelany replied:
Ah that makes sense - yes i was using the parcel in a workspace.

dandelany commented on e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/BindingsTests.java:
```diff
@@ -1898,6 +1900,1677 @@ void conflictedIfDestinationExists() {
+    @Nested
+    class BulkWSRoutes {
+      @Nested
+      class BulkPut {
```
Small/nonblocker, but it would help with readability if we could standardize naming conventions and name handlers (& their associated helpers) for what they *do* rather than the HTTP verb. Eg. we say `BulkPut` in some places and `BulkCreate` in others, same with `BulkPost` vs. `BulkMoveCopy`

Mythicaeda replied:
The reason the handlers were named for their HTTP verb is because they're handling ALL requests of that type. For example, we currently support two types of `POST` requests: move and copy. Once we know which of these requests we're handling, we pass it to a method that's named for the specific action its doing (ie `handleMove`). 

I'd prefer to keep the direct handlers named according to what endpoint they're handling and pull out the uploading/deleting code into sub-methods that the handlers call (ie make a `handleUpload` method that `put` and `bulkPut` called).

dandelany replied:
> they're handling ALL requests of that type

But they're not - they're handling all requests of that type *to* a particular endpoint, which is not really described by the function name. Eg. the function just called `post` doesn't tell me anything about what's being posted/where & doesn't scale well to multiple endpoints. `bulkPost` is better but my preference remains something descriptive, eg. `BulkMoveCopy`, it just gives readers more information when looking at the endpoint -> handler bindings. In Express I often reserve the prefix `handle` specifically for direct handlers to distinguish them from the sub-methods/helpers doing the actual work

Either way, we have an inconsistent mix today, so it would be good to land on some convention here (though ultimately minor!):
```
path("/ws/bulk/{workspaceId}", () -> {
  ApiBuilder.put(this::bulkUpload);
  ApiBuilder.post(this::bulkPost);
  ApiBuilder.delete(this::bulkDelete);
});
path("/ws/{workspaceId}/<path>",
     () -> {
       ApiBuilder.get(this::get);
       ApiBuilder.put(this::put);
       ApiBuilder.delete(this::delete);
       ApiBuilder.post(this::post);
});
path("/ws/{workspaceId}", () -> {
  ApiBuilder.get(this::listWorkspaceContents);
  ApiBuilder.delete(this::deleteWorkspace);
});
```

Mythicaeda replied:
I can rename `put` to `singleItemUpload` and so on, though I'd like to argue in favor of keeping the word `post` in the `post` handlers, as 1) I find `MoveCopy` clunkier and 2) I would not be surprised if we support more types of POST requests as workspaces matures.

Mythicaeda replied:
Addressed in this commit, though if you have different names you'd prefer than please say so and I'll update them: [Rename handler methods](https://github.com/NASA-AMMOS/aerie/pull/1760/commits/1a76ccbce6a3a9fb0183383cff1025ef1bd223d1)

dandelany commented on workspace-server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -44,6 +47,65 @@ Path resolveSubPath(final Path rootPath, final Path filePath) {
+        throw new WorkspaceFileOpException("Path segment '"+ segment+ "' contains reserved name: "+name);
+      }
+    }
+  }
```
Thanks for catching these cases! My only nonblocking architectural complaint here is that devs always have to remember to call *both* resolve/validate functions in order to get a safe path to use, ie.
```
final var path = resolveSubPath(repoPath, directoryPath);
validatePath(path);
```
Feels like it would be safer to bake validation into `resolveSubPath` or some parse function which calls both?

Mythicaeda replied:
Lemme see about making a separate function to call both. The reason it's not baked into `resolveSubPath` is because we only need to validate the path when we're going to try and write to it (at least, as far as I'm aware), but we need to resolve sub paths way more often (ie checking if a file exists, check whether its a directory, etc).

dandelany replied:
Yep fair, didn't think of that subtlety. Thanks.

Mythicaeda replied:
Added a new commit to address this: [Create method to resolve a path for writing and validate it](https://github.com/NASA-AMMOS/aerie/pull/1760/commits/6290a9ada4fe72f944b19da2e389edd1e74d4d31)

dandelany reviewed (approved):
Thanks @Mythicaeda ! Looks good, and I appreciate the additional e2e test coverage on these endpoints. All of my local testing worked as expected, added some notes on potential improvements below but none of them are blockers. This is OK to merge, or lmk if you push any changes for my comments & I'll take another look.

Mythicaeda committed:
Use non-500 handler for Javalin HTTPResponseExceptions

Mythicaeda committed:
Create method to resolve a path for writing and validate it

Mythicaeda committed:
Log when a 500 is returned by the server

Mythicaeda committed:
Extract Duplicate Logic for Upload and Delete

- rename MoveCopyResult to more generic HandlerResult

Mythicaeda committed:
Rename handler methods

Mythicaeda committed:
Properly extract JsonString results

toString leaves behind an extra pair of double quotes within the string, as opposed to returning the contents of the string.

dandelany commented:
👍 reviewed final round of changes with @AaronPlave, this is good to go. Merging.
